src/Servers/Kestrel/shared/test/StreamFillBufferExtensions.cs

```diff
@@ -0,0 +1,59 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Threading;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace System.IO
+{
+    public static class StreamFillBufferExtensions
+    {
+        public static async Task<int> ReadUntilEndAsync(this Stream stream, byte[] buffer, CancellationToken cancellationToken = default)
+        {
+            var offset = 0;
+            int read;
+
+            do
+            {
+                read = await stream.ReadAsync(buffer, offset, buffer.Length - offset, cancellationToken);
+                offset += read;
+            } while (read != 0 && offset < buffer.Length);
+
+            Assert.Equal(0, await stream.ReadAsync(new byte[1], 0, 1, cancellationToken));
+
+            return offset;
+        }
+
+        public static async Task FillEntireBufferAsync(this Stream stream, byte[] buffer, CancellationToken cancellationToken = default)
+        {
+            var offset = 0;
+            int read;
+
+            do
+            {
+                read = await stream.ReadAsync(buffer, offset, buffer.Length - offset, cancellationToken);
+                offset += read;
+            } while (read != 0 && offset < buffer.Length);
+
+            Assert.Equal(buffer.Length, offset);
+        }
+
+        public static async Task<int> ReadAtLeastLengthAsync(this Stream stream, int length, byte[] buffer, CancellationToken cancellationToken = default)
+        {
```

Comment:
nit: this one can share code with FillEntireBufferAsync if we use a sliced buffer if we change to use `Memory<byte>`